import grpc
import json
import logging
import sys
from pathlib import Path

# ğŸ‡¬ğŸ‡§ Add the 'api/gen' directory to the path to find the generated gRPC modules
# ğŸ‡«ğŸ‡· Ajoute le dossier 'api/gen' au path pour trouver les modules gRPC gÃ©nÃ©rÃ©s
api_gen_path = Path(__file__).resolve().parent / 'api/gen'
sys.path.append(str(api_gen_path))

import qiria_pb2
import qiria_pb2_grpc

def run():
    """
    ğŸ‡¬ğŸ‡§ Connects to the Go gRPC server and sends a sample ReportRequest.
    ğŸ‡«ğŸ‡· Se connecte au serveur gRPC Go et envoie une requÃªte d'exemple ReportRequest.
    """
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

    # ğŸ‡¬ğŸ‡§ This is a placeholder token. In a real test, this should be obtained from an auth service.
    # ğŸ‡«ğŸ‡· Ceci est un token de remplacement. Dans un test rÃ©el, il devrait Ãªtre obtenu d'un service d'authentification.
    JWT_TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IlRlc3QgVXNlciIsImlhdCI6MTUxNjIzOTAyMn0.dummy_signature"

    server_address = 'localhost:50051'
    ca_cert_path = 'certs/ca.pem'

    logging.info(f"ğŸ‡¬ğŸ‡§ Attempting to connect to gRPC server at {server_address} using TLS... / ğŸ‡«ğŸ‡· Tentative de connexion au serveur gRPC Ã  {server_address} avec TLS...")

    try:
        # ğŸ‡¬ğŸ‡§ Create a secure channel with TLS, using the CA certificate.
        # ğŸ‡«ğŸ‡· CrÃ©er un canal sÃ©curisÃ© avec TLS, en utilisant le certificat de la CA.
        with open(ca_cert_path, 'rb') as f:
            root_certs = f.read()
        credentials = grpc.ssl_channel_credentials(root_certificates=root_certs)
        channel = grpc.secure_channel(server_address, credentials)

        stub = qiria_pb2_grpc.QiriaCoreStub(channel)

        report_params = {
            "target_host": "srv-prod-db-01",
            "scan_profile": "full"
        }
        request = qiria_pb2.ReportRequest(
            report_id="nis2_compliance_v1",
            parameters_json=json.dumps(report_params)
        )

        # ğŸ‡¬ğŸ‡§ Prepare metadata for authentication.
        # ğŸ‡«ğŸ‡· PrÃ©parer les mÃ©tadonnÃ©es pour l'authentification.
        metadata = [('authorization', f'Bearer {JWT_TOKEN}')]

        logging.info(f"ğŸ‡¬ğŸ‡§ Sending request for report '{request.report_id}'... / ğŸ‡«ğŸ‡· Envoi de la requÃªte pour le rapport '{request.report_id}'...")
        response = stub.RequestReport(request, metadata=metadata)
        logging.info(f"ğŸ‡¬ğŸ‡§ Server responded with Task ID: {response.task_id} and Status: {response.status} / ğŸ‡«ğŸ‡· RÃ©ponse du serveur - Task ID: {response.task_id}, Statut: {response.status}")

    except FileNotFoundError:
        logging.error(f"ğŸ‡¬ğŸ‡§ CA certificate file not found at '{ca_cert_path}'. / ğŸ‡«ğŸ‡· Fichier du certificat CA non trouvÃ© Ã  '{ca_cert_path}'.")
    except grpc.RpcError as e:
        logging.error(f"ğŸ‡¬ğŸ‡§ gRPC error: {e.details()} (code: {e.code().name}) / ğŸ‡«ğŸ‡· Erreur gRPC: {e.details()} (code: {e.code().name})")

if __name__ == '__main__':
    run()